File: ocaml/src/sekreto.ml
(* sekreto: one interface for secrets, wherever they live.

   A Sekreto is an ordered chain of providers. `get` asks each in turn and
   returns the first hit, so an app can be configured from environment
   variables in development and a vault in production without changing a
   line of its own code.

   A provider answers one question: "do you have this secret?" It answers
   with the value, or None to mean "ask the next one". Nothing else about a
   provider is visible to the caller - which is the point: an app reads
   `api.token` and never learns whether it came from the environment, a
   .env file, HashiCorp Vault or a boru vault.

   THE CHAIN IS A voxgig/plugin HOST. Each spec'd provider is an instance
   on it, addressed by name and tag - `hashicorp` for a store named after
   its kind, `hashicorp$prod` otherwise - so `Sekreto.host` reads like the
   chain. Only four kinds are built in: `env`, `memory`, `dotenv` and
   `file`, the ones that read at most a local file. Every kind that opens a
   socket, signs a request or spawns a process lives under `plugins/`, and
   a chain may name one only if the calling project handed it to
   `~plugins`. That is what keeps a chain of built-ins free of TLS, of AWS
   request signing and of seven HTTP vault clients.

   A port of typescript/src/Sekreto.ts, which is canonical. *)

module V = Value
open Secret

(* ---- the core surface, under its canonical names ---------------------

   `Secret` and `Provider` are separate modules only because OCaml compiles
   a module before anything that uses it: the facade needs the four
   built-in kinds, which need the name functions and the error. A caller
   writes `Sekreto.envkey` and never has to know. *)

exception Sekreto_error = Secret.Sekreto_error

type provider = Secret.provider = {
  lookup : string -> string option;
  describe : unit -> string;
}

type vaultref = Secret.vaultref = { path : string; field : string }

let fail = Secret.fail
let dropsuffix = Secret.dropsuffix
let segments = Secret.segments
let validsegment = Secret.validsegment
let validname = Secret.validname
let checkname = Secret.checkname
let upper = Secret.upper
let envkey = Secret.envkey
let vaultref = Secret.vaultref
let flatname = Secret.flatname
let awsparam = Secret.awsparam
let unescape = Secret.unescape
let trim = Secret.trim
let assocset = Secret.assocset
let parsedotenv = Secret.parsedotenv
let replaceall = Secret.replaceall
let redact = Secret.redact
let storename = Secret.storename
let findsub = Secret.findsub
let splitfirst = Secret.splitfirst

(* A plugin error carries a parseable message and no printer of its own, so
   a host that catches one - the CLI, the conformance runner - would read
   `Types.Plugin_error(_)` instead of the diagnostic. Registered here
   because this is the module every consumer already links. *)
let () =
  Printexc.register_printer (function
    | Types.Plugin_error err -> Some err.Types.message
    | _ -> None)

(* ---- refusals that name the fix -------------------------------------- *)

(* The message for a kind the catalog does not hold.

   A kind sekreto has never heard of is a typo; a kind that exists as a
   plugin but was not passed in is the split working as designed and
   telling you what to pass. Collapsing the two was the first thing that
   made the split confusing to use. *)
let unknownkind (kind : string) (catalog : Defs.catalog) : string =
  let available =
    String.concat ", " (List.map V.as_str (V.items (Catalog.names catalog)))
  in
  "sekreto: unknown provider kind: " ^ kind ^ " (available: " ^ available ^ ")"
  ^
  if List.mem kind Provider.pluginkinds then
    " - " ^ kind ^ " is a sekreto plugin, not built in: pass it in the plugins option"
  else ""

(* A Sekreto_error that crossed the plugin boundary comes back out as
   itself, byte for byte. Anything else is not sekreto's to rewrite: it
   surfaces as the host reports it, naming the instance. *)
let unwrap (err : exn) : exn =
  match err with
  | Types.Plugin_error report
    when Provider.error_code = report.Types.code
         && V.is_str (V.get report.Types.details "cause") ->
    Sekreto_error (V.as_str (V.get report.Types.details "cause"))
  | other -> other

(* ---- the facade ------------------------------------------------------ *)

(* One provider in the chain, under the store name it answers to, and the
   ref of the plugin instance that built it - `""` for a live provider
   handed in directly, which no instance backs. *)
type entry = { store : string; eref : string; eprovider : provider }

type cached = { cstore : string; cname : string; cvalue : string }

type t = {
  (* The voxgig/plugin host every spec'd provider is an instance of. Read
     it for introspection - `Host.list` names each instance and its status
     - and nothing on it advances the chain. *)
  thost : Defs.host;
  (* The definitions this Sekreto can build: the built-ins, then whatever
     `~plugins` handed in. *)
  tcatalog : Defs.catalog;
  mutable entries : entry list;
  (* A list, not a map: the store a value came from stays attached, and
     redaction order does not vary between runs. *)
  mutable cache : cached list;
  (* Every value ever resolved, for `redacttext`. Kept independently of the
     read cache so that redaction still works when caching is off -
     otherwise an uncached Sekreto would silently disable redaction and leak
     secrets to logs. Append-only for the object's life: neither `refresh`
     nor `close` clears it. Held reversed, and reversed back on read. *)
  mutable seen : string list;
  docache : bool;
}

let host (self : t) : Defs.host = self.thost
let catalog (self : t) : Defs.catalog = self.tcatalog

(* Built-ins first, then the plugins, into one catalog: a plugin that names
   a built-in kind replaces it, which is how a host substitutes an
   implementation and never an accident, because the four names are
   documented. *)
let makecatalog (plugins : Defs.definition list) : Defs.catalog =
  let catalog = Catalog.makecatalog () in
  List.iter (Catalog.add catalog) (Provider.builtins () @ plugins);
  catalog

let empty (plugins : Defs.definition list) (cache : bool) : t =
  let catalog = makecatalog plugins in
  {
    thost = Host.makehost { Defs.nohostoptions with ocatalog = Some catalog };
    tcatalog = catalog;
    entries = [];
    cache = [];
    seen = [];
    docache = cache;
  }

(* One chain entry, as a plugin instance.

   The instance is `kind` for a store named after its kind and `kind$store`
   otherwise, so the host reads like the chain. A store name already taken
   gets a numbered tag from the host instead, because two providers MAY
   share a store name - a directed read walks both - and an instance ref
   may not. *)
let declare (self : t) (spec : Provider.spec) : entry =
  let kind = spec.Provider.kind in

  if not (Catalog.has self.tcatalog kind) then fail (unknownkind kind self.tcatalog);

  let store = if "" = spec.Provider.name then kind else spec.Provider.name in

  if not (Ref.checktag (V.vstr store)) then fail ("sekreto: invalid store name: " ^ store);

  let wanted = if store = kind then kind else Ref.formatref (V.vstr kind) (V.vstr store) in
  let iref =
    if None = Host.instance self.thost wanted then wanted else Host.autotag self.thost kind
  in

  (* `load` runs the definition's `define`, which builds the provider from
     the spec; `activate` takes the instance live. Nothing is contacted by
     either: a provider opens nothing until its first lookup. *)
  (try
     ignore (Host.load self.thost iref { Defs.nospec with soptions = Some (Provider.optionsof spec) });
     ignore (Host.activate self.thost iref)
   with err -> raise (unwrap err));

  let handle =
    match Host.exports self.thost (iref ^ "/" ^ Provider.provider_export) with
    | Some held when V.is_num held -> int_of_float (V.as_num held)
    | _ -> 0
  in

  match Provider.providerof handle with
  | Some made -> { store; eref = iref; eprovider = made }
  (* A definition that loads, activates and exports no provider is not a
     provider plugin at all, and saying so by name is more use than a
     missing-export error from the host. *)
  | None -> fail ("sekreto: plugin " ^ kind ^ " exported no provider")

(* Build a chain from live providers. `names` is positional; an entry left
   empty falls back to the provider's kind. No plugin instance backs any of
   them, so the host stays empty. Construction contacts nothing - the first
   network call is the first lookup. *)
let make ?(names = []) ?(cache = true) (providers : provider list) : t =
  let self = empty [] cache in
  self.entries <-
    List.mapi
      (fun index made ->
        let named = match List.nth_opt names index with Some n -> n | None -> "" in
        { store = (if "" <> named then named else storename made); eref = ""; eprovider = made })
      providers;
  self

(* Build a chain from declarative provider specs - the same shape the
   shared spec and an app's config file use.

   `~plugins` is the whole loading mechanism: static, explicit, and a list
   handed to a constructor rather than a side effect of importing. A kind
   that is not built in and was not passed here cannot be built, and the
   refusal says so. *)
let sekreto ?(cache = true) ?(plugins = []) (specs : Provider.spec list) : t =
  let self = empty plugins cache in

  match List.map (declare self) specs with
  | entries ->
    self.entries <- entries;
    self
  | exception err ->
    (* Whatever was declared before the refusal is torn down, so a chain
       that never came into being leaves no live instance and no stashed
       provider behind. Each definition's `close` hands its own handle
       back, so nothing here has to know which were taken. *)
    (try Host.close self.thost with _ -> ());
    raise err

(* The single path both readers share. The name is validated FIRST, before
   the cache and before the first provider is asked. *)
let resolve (self : t) (store : string) (name : string) (useentries : entry list) :
    string option =
  ignore (checkname name);

  let hit =
    if self.docache then
      List.find_opt (fun c -> store = c.cstore && name = c.cname) self.cache
    else None
  in

  match hit with
  | Some c -> Some c.cvalue
  | None ->
    let rec walk = function
      | [] -> None
      | e :: rest -> (
        (* The empty string is a hit; None and only None is a miss. *)
        match e.eprovider.lookup name with Some value -> Some value | None -> walk rest)
    in
    let found = walk useentries in
    (match found with
    | Some value ->
      if self.docache then self.cache <- self.cache @ [ { cstore = store; cname = name; cvalue = value } ];
      self.seen <- value :: self.seen
    (* Misses are never cached. *)
    | None -> ());
    found

(* The secret, or None if no provider has it. Named `tryget` because `try`
   is an OCaml keyword. *)
let tryget (self : t) (name : string) : string option = resolve self "" name self.entries

(* The secret, or a Sekreto_error if no provider has it. *)
let get (self : t) (name : string) : string =
  match tryget self name with
  | Some value -> value
  | None -> fail ("sekreto: unknown secret: " ^ name)

(* The secret from one named store, or None if that store does not have it.

   Naming a store that is not in the chain is an error, not a miss:
   `tryget` already means "this store may not have it", so it cannot also
   mean "this store may not exist" without hiding a typo. The refusal comes
   before the name is validated. *)
let tryfrom (self : t) (store : string) (name : string) : string option =
  let matching = List.filter (fun e -> store = e.store) self.entries in
  if [] = matching then fail ("sekreto: unknown store: " ^ store);
  resolve self store name matching

let getfrom (self : t) (store : string) (name : string) : string =
  match tryfrom self store name with
  | Some value -> value
  | None -> fail ("sekreto: unknown secret: " ^ store ^ ":" ^ name)

let has (self : t) (name : string) : bool = None <> tryget self name
let hasin (self : t) (store : string) (name : string) : bool = None <> tryfrom self store name

(* Every named secret at once. Missing ones are an error. *)
let all (self : t) (names : string list) : (string * string) list =
  List.map (fun name -> (name, get self name)) names

(* A description of each provider, in resolution order, repeats kept. *)
let sources (self : t) : string list = List.map (fun e -> e.eprovider.describe ()) self.entries

(* The name of each store `getfrom` can address, in resolution order and
   without repeats. *)
let stores (self : t) : string list =
  let rec dedupe seen = function
    | [] -> []
    | head :: rest ->
      if List.mem head seen then dedupe seen rest else head :: dedupe (head :: seen) rest
  in
  dedupe [] (List.map (fun e -> e.store) self.entries)

(* Replace every value this Sekreto has resolved with `[redacted]`. Works
   whether or not caching is on. *)
let redacttext (self : t) (text : string) : string = redact text (List.rev self.seen)

(* Drop cached values, so the next `get` asks the providers again. The
   redaction history is not touched. *)
let refresh (self : t) : unit = self.cache <- []

(* Tear the chain down: every plugin instance is deactivated and unloaded,
   in reverse, releasing whatever a provider acquired at activation.
   Afterwards `stores` and `sources` are empty, `get` raises and `tryget`
   misses - and redaction still knows every value ever resolved, which is
   the whole point of keeping `seen` separately. *)
let close (self : t) : unit =
  Host.close self.thost;
  self.entries <- [];
  self.cache <- []

(* ---- print hooks -----------------------------------------------------

   `cache` and `seen` are ordinary fields, so the default printer of a
   `t` would put every resolved secret on the page. Neither hook below can
   reach a value. *)

let to_string (self : t) : string = "Sekreto { stores: [ " ^ String.concat ", " (stores self) ^ " ] }"
let to_json (self : t) : string =
  (* Built as a Json value and written by the port's own writer, NOT
     concatenated. Canonical returns an OBJECT here
     (typescript/src/Sekreto.ts toJSON) and lets JSON.stringify escape it,
     so a store name carrying a quote, a backslash or a control character
     is valid JSON in every port. Assembling the text by hand produced
     invalid JSON for exactly those names, and no spec entry covers this
     hook -- an audit caught it, not the corpus. *)
  Json.stringify (Json.obj [ ("stores", Json.arr (List.map Json.str (stores self))) ])

File: lua/src/sekreto.lua
-- sekreto: one interface for secrets, wherever they live.
--
-- A Sekreto is an ordered chain of providers. `get` asks each in turn and
-- returns the first hit, so an app can be configured from environment
-- variables in development and a vault in production without changing a
-- line of its own code.
--
-- This module is the whole core surface: the facade, the pure name
-- functions (re-exported from sekreto.name, which the providers need
-- too), the error type, and the four built-in provider kinds.
--
-- THE CORE REQUIRES NO PROVIDER THAT OPENS A SOCKET, SPAWNS A PROCESS OR
-- SIGNS A REQUEST. The four built-in kinds - env, memory, dotenv, file -
-- read at most a local file; every other kind is a voxgig/plugin
-- definition under sekreto/plugins/, and a chain may name one only if the
-- calling project handed it in through `plugins`:
--
--     local sekreto = require('sekreto')
--     local hashicorp = require('sekreto.plugins.hashicorp').hashicorp
--
--     local secrets = sekreto.sekreto({
--       plugins = { hashicorp },
--       providers = {
--         { kind = 'env' },
--         { kind = 'hashicorp', addr = addr, token = token },
--       },
--     })
--
-- or, for every kind at once, `allplugins` from `sekreto.plugins`. See
-- docs/design/plugin-providers.md.
--
-- A port of typescript/src/Sekreto.ts, which is canonical.

local plugin = require('plugin')

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')

local M = {}

-- ---------------------------------------------------------- re-exports

M.SekretoError = err.SekretoError
M.issekretoerror = err.issekretoerror
M.errmessage = err.message

M.validname = name.validname
M.checkname = name.checkname
M.envkey = name.envkey
M.vaultref = name.vaultref
M.flatname = name.flatname
M.awsparam = name.awsparam
M.parsedotenv = name.parsedotenv
M.redact = name.redact
M.storename = name.storename

M.checkaddr = addr.checkaddr
M.safeaddr = addr.safeaddr

M.spec = providers.spec
M.authspec = providers.authspec

-- The plugin bridge, and the four kinds built on it. `providerplugin` is
-- how a calling project adds a fifth.
M.providerplugin = providers.providerplugin
M.BUILTINS = providers.BUILTINS
M.KINDS = providers.KINDS
M.PROVIDER_EXPORT = providers.PROVIDER_EXPORT
M.ERROR_CODE = providers.ERROR_CODE

-- ------------------------------------------------------------- the type

local Sekreto = {}
Sekreto.__index = Sekreto

--- The stores this Sekreto can be asked by name, without repeats.
--- Defined before the print hooks, which are the only reason it needs to
--- come first.
function Sekreto:stores()
  local out = {}
  local seen = {}

  for _, entry in ipairs(self.entries) do
    if nil == seen[entry.store] then
      seen[entry.store] = true
      out[#out + 1] = entry.store
    end
  end

  return out
end

-- Print hooks, and they are not optional. `cache` and `seen` are
-- ordinary fields, so anything that walks or prints a Sekreto - a
-- debugger, a logging helper, an error formatter - would otherwise emit
-- every secret ever resolved. Neither hook reaches a value.
Sekreto.__tostring = function(self)
  return 'Sekreto { stores: [ ' .. table.concat(self:stores(), ', ') .. ' ] }'
end
Sekreto.__name = 'Sekreto'

--- The JSON-shaped form: the store names, and nothing else.
function Sekreto:tojson()
  return { stores = self:stores() }
end

-- ------------------------------------------------------- the plugin seam

--- Is this a voxgig/plugin definition?
local function isdefinition(value)
  return 'table' == type(value) and 'string' == type(value.name)
    and 'function' == type(value.define)
end

--- The name a table is loaded under, when it is a module.
---
--- Lua gives a module no identity of its own, so the registry is asked.
--- It is only ever consulted on the error path below, where naming the
--- module is the whole point of the message.
local function modulename(value)
  for key, loaded in pairs(package.loaded) do
    if loaded == value then
      return key
    end
  end
  return nil
end

--- A plugin entry, checked to be a definition before the catalog sees it.
---
--- `require('sekreto.plugins.hashicorp')` hands back the MODULE, and the
--- definition is one field further on - so the thing nearest to hand is
--- not a definition, and a module in the catalog would fail deep inside
--- voxgig/plugin with a message about a definition name. Refused here
--- instead, naming the module and what to pass out of it.
local function definition(entry)
  if isdefinition(entry) then
    return entry
  end

  local modname = ('table' == type(entry)) and modulename(entry) or nil

  if nil ~= modname then
    -- What the module holds that could have been meant: a definition, or
    -- a list of them.
    local holds = {}
    for key, value in pairs(entry) do
      if isdefinition(value) or
        ('table' == type(value) and isdefinition(value[1]))
      then
        holds[#holds + 1] = key
      end
    end
    table.sort(holds)

    if 0 < #holds then
      err.fail('sekreto: not a plugin definition: the module ' .. modname ..
        " - pass a definition it holds, such as require('" .. modname ..
        "')." .. holds[1])
    end

    err.fail('sekreto: not a plugin definition: the module ' .. modname)
  end

  err.fail('sekreto: not a plugin definition: ' .. tostring(entry))
end

--- The message for a kind the catalog does not hold.
---
--- A kind sekreto has never heard of is a typo; a kind that exists as a
--- plugin but was not passed in is the split working as designed and
--- telling you what to pass. Collapsing the two was the first thing that
--- made the split confusing to use.
local function unknownkind(kind, catalog)
  local message = 'sekreto: unknown provider kind: ' .. tostring(kind) ..
    ' (available: ' .. table.concat(catalog:names(), ', ') .. ')'

  for _, known in ipairs(providers.KINDS.plugin) do
    if known == kind then
      return message .. ' - ' .. tostring(kind) ..
        ' is a sekreto plugin, not built in: pass it in the plugins option'
    end
  end

  return message
end

--- A SekretoError that crossed the plugin boundary comes back out as
--- itself, byte for byte. Anything else is not sekreto's to rewrite.
local function unwrap(raised)
  if providers.ERROR_CODE == plugin.codeof(raised) then
    local cause = ('table' == type(raised.details)) and raised.details.cause or nil
    if 'string' == type(cause) then
      return err.SekretoError(cause)
    end
  end

  return raised
end

--- One chain entry, as a plugin instance.
---
--- The instance is `kind` for a store named after its kind and
--- `kind$store` otherwise - `hashicorp$prod` - so `host:list()` reads like
--- the chain. A store name that is already taken gets a numbered tag from
--- the host instead, because two providers MAY share a store name (a
--- directed read walks both) and an instance ref may not.
function Sekreto:declare(spec)
  local kind = ('table' == type(spec)) and spec.kind or nil

  if 'string' ~= type(kind) or not self.catalog:has(kind) then
    err.fail(unknownkind(kind, self.catalog))
  end

  local store = providers.nonempty(spec.name) and spec.name or kind

  if 'string' ~= type(store) or not plugin.check_tag(store) then
    err.fail('sekreto: invalid store name: ' .. tostring(store))
  end

  local ref = (store == kind) and kind or plugin.format_ref(kind, store)
  if nil ~= self.host:instance(ref) then
    ref = self.host:autotag(kind)
  end

  -- `load` runs the definition's `define`, which builds the provider from
  -- the spec; `activate` takes the instance live. Nothing is contacted by
  -- either: a provider opens nothing until its first lookup.
  local ok, raised = pcall(function()
    self.host:load(ref, plugin.types.map({ options = spec }))
    self.host:activate(ref)
  end)

  if not ok then
    error(unwrap(raised), 0)
  end

  return {
    store = store,
    provider = self.host:exports(ref .. '/' .. providers.PROVIDER_EXPORT),
  }
end

--- Make a Sekreto.
---
--- `options` carries `plugins` (the definitions this chain may build,
--- beyond the four built-ins), `providers` (the chain itself, as specs or
--- as live providers) and `cache`. Construction contacts nothing: the
--- first network call is the first lookup.
function M.Sekreto(options)
  local self = setmetatable({}, Sekreto)
  local opts = options or {}

  -- Built-ins first, then the plugins, into one catalog: a plugin that
  -- names a built-in kind replaces it, which is how a host substitutes an
  -- implementation and never an accident, because the four names are
  -- documented.
  --
  -- `catalog` is the definitions this Sekreto can build; `host` is the
  -- voxgig/plugin host every spec'd provider is an instance of. Read them
  -- for introspection - `host:list()` names each store's ref and status -
  -- and nothing on either advances the chain.
  local defs = {}
  for _, builtin in ipairs(providers.BUILTINS) do
    defs[#defs + 1] = builtin
  end
  for _, given in ipairs(opts.plugins or {}) do
    defs[#defs + 1] = definition(given)
  end

  self.catalog = plugin.make_catalog(defs)
  self.host = plugin.make_host(plugin.types.map({ catalog = self.catalog }))

  -- {store, provider} pairs, in chain order. A provider handed in live is
  -- backed by no instance; a spec'd one is an instance of its kind on the
  -- host.
  self.entries = {}

  for index, entry in ipairs(opts.providers or {}) do
    if 'function' == type(entry.lookup) then
      self.entries[index] = { store = name.storename(entry), provider = entry }
    else
      self.entries[index] = self:declare(entry)
    end
  end

  -- A list, not a map: the store a value came from stays attached, and
  -- redaction order does not vary between runs.
  self.cache = {}

  -- Every value ever resolved, for redact(). Kept independently of the
  -- read cache so that redaction still works when caching is off -
  -- otherwise an uncached Sekreto would silently disable redact() and
  -- leak secrets to logs. Append-only for the object's life: neither
  -- refresh() nor close() clears it.
  self.seen = {}

  -- A strict identity test, as canonical: only an exact `false` turns
  -- caching off, so nil leaves it on.
  self.docache = (false ~= opts.cache)

  return self
end

--- The one resolution path both readers share.
function Sekreto:resolve(store, secret, useentries)
  -- The name is validated FIRST: before the cache, before the first
  -- provider.
  name.checkname(secret)

  if self.docache then
    for _, hit in ipairs(self.cache) do
      if store == hit.store and secret == hit.name then
        return hit.value
      end
    end
  end

  for _, entry in ipairs(useentries) do
    local found = entry.provider.lookup(secret)

    -- The empty string is a HIT. Only nil is a miss, and a provider that
    -- raises is not caught here: the error propagates out of get/try.
    if nil ~= found then
      if self.docache then
        self.cache[#self.cache + 1] = { store = store, name = secret, value = found }
      end
      self.seen[#self.seen + 1] = found
      return found
    end
  end

  -- Misses are never cached.
  return nil
end

--- The secret, or nil if no provider has it.
---
--- Named `tryget` because `try` is not usable as a Lua field name without
--- quoting at every call site; `sekreto['try']` is kept as an alias for
--- callers translating from canonical.
function Sekreto:tryget(secret)
  return self:resolve('', secret, self.entries)
end

Sekreto.try = Sekreto.tryget

--- The secret, or a SekretoError if no provider has it.
function Sekreto:get(secret)
  local found = self:tryget(secret)

  if nil == found then
    err.fail('sekreto: unknown secret: ' .. tostring(secret))
  end

  return found
end

--- The secret from one named store, or nil if that store does not have
--- it.
---
--- Naming a store that is not in the chain is an error, not a miss:
--- `tryget` already means "this store may not have it", so it cannot also
--- mean "this store may not exist" without hiding a typo. It is raised
--- BEFORE the name is validated.
function Sekreto:tryfrom(store, secret)
  local matching = {}

  for _, entry in ipairs(self.entries) do
    if store == entry.store then
      matching[#matching + 1] = entry
    end
  end

  if 0 == #matching then
    err.fail('sekreto: unknown store: ' .. tostring(store))
  end

  return self:resolve(store, secret, matching)
end

--- The secret from one named store, or a SekretoError.
function Sekreto:getfrom(store, secret)
  local found = self:tryfrom(store, secret)

  if nil == found then
    err.fail('sekreto: unknown secret: ' .. tostring(store) .. ':' .. tostring(secret))
  end

  return found
end

--- Does any provider have this secret?
function Sekreto:has(secret)
  return nil ~= self:tryget(secret)
end

--- Does this named store have this secret?
function Sekreto:hasin(store, secret)
  return nil ~= self:tryfrom(store, secret)
end

--- Every named secret at once. Missing ones are an error.
function Sekreto:all(secrets)
  local out = {}
  local order = {}

  for _, secret in ipairs(secrets) do
    out[secret] = self:get(secret)
    order[#order + 1] = secret
  end

  return out, order
end

--- A description of each provider, in resolution order, repeats kept.
function Sekreto:sources()
  local out = {}

  for index, entry in ipairs(self.entries) do
    out[index] = entry.provider.describe()
  end

  return out
end

--- Replace every value this Sekreto has resolved with `[redacted]`.
--- Works whether or not caching is enabled.
function Sekreto:redact(text)
  return name.redact(text, self.seen)
end

--- Drop cached values, so the next `get` asks the providers again. The
--- redaction history is NOT cleared.
function Sekreto:refresh()
  self.cache = {}
end

--- Tear the chain down: every plugin instance is deactivated and
--- unloaded, in reverse, releasing whatever a provider acquired at
--- activation. Afterwards `stores()` and `sources()` are empty, `tryget`
--- misses and `get` raises - and `redact` still knows every value ever
--- resolved.
function Sekreto:close()
  self.host:close()
  self.entries = {}
  self.cache = {}
end

M.SekretoClass = Sekreto

-- ---------------------------------------------------------- the factory

--- Make a Sekreto from options - the same shape the shared spec and an
--- app's config file use.
---
--- A `providers` element that already has a callable `lookup` is taken as
--- a live provider (duck-typed, never by class identity) and keeps its
--- own describe()-derived store name.
function M.sekreto(options)
  return M.Sekreto(options)
end

return M
